src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -65,52 +67,272 @@ ucs_status_t uct_rdmacm_cm_reject(struct rdma_cm_id *id)
     return UCS_OK;
 }
 
-ucs_status_t uct_rdmacm_cm_get_cq(uct_rdmacm_cm_t *cm, struct ibv_context *verbs,
-                                  struct ibv_cq **cq_p)
+static ucs_status_t
+uct_rdmacm_cm_query_reserved_qpn_cap(struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_qp_context *ctx)
+{
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_out)] = {};
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(query_hca_cap_in)]   = {};
+    void *cap;
+    int ret;
+
+    if (!mlx5dv_is_supported(verbs->device)) {
+        ucs_debug("mlx5dv is not supported");
+        goto err_out;
+    }
+
+    cap = UCT_IB_MLX5DV_ADDR_OF(query_hca_cap_out, out, capability);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, opcode,
+                      UCT_IB_MLX5_CMD_OP_QUERY_HCA_CAP);
+    UCT_IB_MLX5DV_SET(query_hca_cap_in, in, op_mod,
+                      (UCT_IB_MLX5_CAP_GENERAL << 1) |
+                      UCT_IB_MLX5_HCA_CAP_OPMOD_GET_CUR);
+    ret = mlx5dv_devx_general_cmd(verbs, in, sizeof(in),
+                                  out, sizeof(out));
+    if (ret != 0) {
+        ucs_debug("mlx5dv_devx_general_cmd(QUERY_HCA_CAP) failed: %m");
+        goto err_out;
+    }
+
+    ctx->reserved_qpn_ctx.general_obj_types_caps =
+            UCT_IB_MLX5DV_GET64(cmd_hca_cap, cap, general_obj_types);
+    if (!(ctx->reserved_qpn_ctx.general_obj_types_caps &
+        UCS_BIT(UCT_IB_MLX5_OBJ_TYPE_RESERVED_QPN))) {
+        ucs_debug("general_obj_types_caps: reserved qpn is not support");
+        goto err_out;
+    }
+
+    ctx->reserved_qpn_support = true;
+    return UCS_OK;
+
+err_out:
+#endif
+    ctx->reserved_qpn_support = false;
+    return UCS_ERR_UNSUPPORTED;
+}
+
+static ucs_status_t
+uct_rdmacm_cm_query_reserved_qpn_cap_2(struct ibv_context *verbs,
```

Comment:
- should return void. the query result is saved already to 'ctx'
- maybe unite with uct_rdmacm_cm_query_reserved_qpn_cap and rename it to uct_rdmacm_cm_device_context_init() - and this function will fill all fields in uct_rdmacm_cm_device_context_t ? it will eliminate some duplicated code

Reply:
Changed to uct_rdmacm_cm_device_context_init(), the return value is used to indicate the status of calling the function now.